unified_planning/engines/interpreted_functions_planner.py

```diff
@@ -0,0 +1,229 @@
+        supported_kind.set_conditions_kind("UNIVERSAL_CONDITIONS")
+
+        supported_kind.set_effects_kind("CONDITIONAL_EFFECTS")
+        supported_kind.set_effects_kind("INCREASE_EFFECTS")
+        supported_kind.set_effects_kind("DECREASE_EFFECTS")
+        supported_kind.set_effects_kind("STATIC_FLUENTS_IN_BOOLEAN_ASSIGNMENTS")
+        supported_kind.set_effects_kind("STATIC_FLUENTS_IN_NUMERIC_ASSIGNMENTS")
+        supported_kind.set_effects_kind("STATIC_FLUENTS_IN_OBJECT_ASSIGNMENTS")
+        supported_kind.set_effects_kind("FLUENTS_IN_BOOLEAN_ASSIGNMENTS")
+        supported_kind.set_effects_kind("FLUENTS_IN_NUMERIC_ASSIGNMENTS")
+        supported_kind.set_effects_kind("FLUENTS_IN_OBJECT_ASSIGNMENTS")
+        supported_kind.set_time("CONTINUOUS_TIME")
+        supported_kind.set_time("DISCRETE_TIME")
+        supported_kind.set_time("INTERMEDIATE_CONDITIONS_AND_EFFECTS")
+        supported_kind.set_time("EXTERNAL_CONDITIONS_AND_EFFECTS")
+        supported_kind.set_time("TIMED_EFFECTS")
+        supported_kind.set_time("TIMED_GOALS")
+        supported_kind.set_time("DURATION_INEQUALITIES")
+        supported_kind.set_time("SELF_OVERLAPPING")
+        supported_kind.set_expression_duration("STATIC_FLUENTS_IN_DURATIONS")
+        supported_kind.set_expression_duration("FLUENTS_IN_DURATIONS")
+        supported_kind.set_expression_duration("INT_TYPE_DURATIONS")
+        supported_kind.set_expression_duration("REAL_TYPE_DURATIONS")
+        supported_kind.set_simulated_entities("SIMULATED_EFFECTS")
+        final_supported_kind = supported_kind.intersection(engine.supported_kind())
+        additive_supported_kind = ProblemKind(version=LATEST_PROBLEM_KIND_VERSION)
+        additive_supported_kind.set_expression_duration(
+            "INTERPRETED_FUNCTIONS_IN_DURATIONS"
+        )
+        additive_supported_kind.set_conditions_kind(
+            "INTERPRETED_FUNCTIONS_IN_CONDITIONS"
+        )
+        additive_supported_kind.set_effects_kind(
+            "INTERPRETED_FUNCTIONS_IN_NUMERIC_ASSIGNMENTS"
+        )
+        additive_supported_kind.set_effects_kind(
+            "INTERPRETED_FUNCTIONS_IN_BOOLEAN_ASSIGNMENTS"
+        )
+        return final_supported_kind.union(additive_supported_kind)
+
+    @staticmethod
+    def _supports(problem_kind: "ProblemKind", engine: Type[Engine]) -> bool:
+        return problem_kind <= InterpretedFunctionsPlanner._supported_kind(engine)
+
+    def _solve(
+        self,
+        problem: "up.model.AbstractProblem",
+        heuristic: Optional[Callable[["up.model.state.State"], Optional[float]]] = None,
+        timeout: Optional[float] = None,
+        output_stream: Optional[IO[str]] = None,
+    ) -> "PlanGenerationResult":
+        assert isinstance(problem, up.model.Problem)
+        assert isinstance(self.engine, mixins.OneshotPlannerMixin)
+        em = problem.environment.expression_manager
+        f = problem.environment.factory
+
```

Comment:
This method can be refactored a little bit. It is not needed to have two pieces of code that call the Compiler.

Something like this:
```python
start = time.time()
knowledge = {}
while True:
    if timeout is not None:
        timeout -= time.time() - start
        if timeout <= 0:
            return PlanGenerationResult(TIMEOUT, None, self.name)
    with InterpretedFunctionsRemover(knowledge) as if_remover:
        comp_res = if_remover.compile(problem)
    res = self.engine.solve(comp_res.problem, heuristic, timeout, output_stream)
    if res.status in up.engines.results.POSITIVE_OUTCOMES:
        plan = res.plan.replace_action_instances(comp_res.map_back_action_instance)
        if plan.kind == PlanKind.SEQUENTIAL_PLAN:
            validator = SequentialPlanValidator()
        elif plan.kind == PlanKind.TIME_TRIGGERED_PLAN:
            validator = TimeTriggeredPlanValidator()
        else:
            raise 
        validation_result = validator.validate(problem, plan)
        if validation_result.status == ValidationResultStatus.VALID:
            return PlanGenerationResult(res.status, plan, self.name, log_messages=res.log_messages)
        else:
            knowledge.update(validation_result.calculated_interpreted_functions)
    else:
        return PlanGenerationResult(res.status, None, self.name)
```